ant/base/ant.py

```diff
@@ -142,18 +147,29 @@ def _worker(self):
                                          Message.ID.RESPONSE_CHANNEL_ID]:
                         self._events.put(('response', (message._data[0],
                                                        message._id, message._data[1:])))
-                    # Response (other)
+ 
+                    # CHANNEL MESSAGE (0x40)
+					# Response (other) # Channel Response, in Channel Message
                     elif (message._id == Message.ID.RESPONSE_CHANNEL
                           and message._data[1] != 0x01):
                         self._events.put(('response', (message._data[0],
                                                        message._data[1], message._data[2:])))
-                    # Channel event
+
+                    # EVENT_TX in Channel Event, in Channel Message				# added to support differentiate if it an TX Event
+                    elif (message._id == Message.ID.RESPONSE_CHANNEL			# added to support differentiate if it an TX Event
+                          and message._data[1] == 0x01							# added to support differentiate if it an TX Event
+                          and message._data[2] == Message.Code.EVENT_TX):		# added to support differentiate if it an TX Event
+                        self._events.put(('event', (message._data[0], 			# EVENT_TX is in data[2]
+                                                       message._data[2], message._data[3:])))# added to support differentiate if it an TX Event
```

Comment:
Could you explain a bit more what this does and why it is needed?

Reply:
I try... ;)
Within MESSAGE_ID 0x40 we can have both:
- a Channel Response (in response to a message) or a Channel Event (generated by an RF event on the ANT device). => Page 54 & 115 on the ANT+ Specification 

The important thing here is that Message data are starting on Byte [3]  not [2]!

I added these line to get 100% sure this is an exact EVENT_TX by checking:
- message._data[1] is 0x01, as everytime given on an EVENT_TX
- and also if the data[2] refers to this Message Code

In your code some lines below you grab the RESPONSE_CHANNEL Message also, and I'm not 100% sure why I overview this, or why you added this in here. BUT if this should refer to EVENT_TX, the putting of (channel, event, data) has to be changed from: message._data[0], message._data[1], message._data[2:])))
TO:
message._data[0], message._data[2], message._data[3:])))
As described Event is on data[2] and DATA starting on data[3]  in case of an EVENT_TX.

Does this makes sence?

Reply:
I've read the relevant sections of the "ANT Message Protocol and Usage" and yes, you are correct. I would change the existing line to (adding the extra `and` just for clarity):
 
```python
                    elif (message._id == Message.ID.RESPONSE_CHANNEL
                          and message._data[1] == 0x01):
                        _logger.debug("Got channel event, %r", message)
                        self._events.put(('event', (message._data[0],
                                                    message._data[2], message._data[3:])))
```

And move it to just under the `# Channel event` comment above so it is closer to the channel response.

Thanks for the clarification :+1: